[Week 1] Submission

glowing713 opened the pull request:
# 1주차 미션 제출 (이선환)

## 접근 방법

> 미션 코드를 구현하기 위해서 어떤 생각들을 했는지 정리해보세요.
- **TDD 방법론**
  - TDD 방법론을 적용하기로 했기 때문에 요구사항을 하나씩 구현할 때마다 다음과 같은 3가지 절차로 접근하게 되었습니다.
  - 실패하는 테스트 코드 작성(최소한의 단위) → 통과하기 위한 최소한의 기능 구현 → 리팩토링
- **프로젝트 구성에 대한 이해**
  - package.json: vue boilerplate 프로젝트가 어떤 라이브러리를 포함하고 있고, 어떤 빌드 스크립트를 사용하는지 확인하였습니다.
    - 테스트 코드 수정사항에 따른 결과를 실시간으로 확인하기 위해 npm 스크립트를 추가(기존 테스트 코드 실행 스크립트에 watchAll 옵션 추가)
  - jsconfig.json: 기존의 세팅이 프로젝트에 필수적인지, 더 추가하거나 제외할 옵션은 없는지 확인
    - typescript, vscode 공식문서를 확인하며 각 옵션을 학습하고 [블로그](https://glowdev.tistory.com/102)에 기록
- **커스텀 컴포넌트**
  - 해당 과제의 경우 하나의 input 엘리먼트, 하나의 p 엘리먼트(또는 div, 등등), 그리고 두 개의 button 엘리먼트를 사용합니다.
  - 재사용성이 높은 컴포넌트를 작성하는 역량을 기르고 싶어서 HTML 엘리먼트의 wrapper 컴포넌트를 작성해보기로 하였습니다.
    - BaseInput, BaseButton 컴포넌트를 구현
    - 각 컴포넌트는 prop을 통해 필요한 값을 전달받게 되고, data는 이들의 부모 컴포넌트인 HelloWorld에서만 처리
    - BaseInput: v-model과 v-bind/v-on 디렉티브를 사용하였고, emit을 통해 input에 따른 값의 업데이트가 발생
    - BaseButton: func prop으로 넘겨준 값에 따라 버튼의 기능이 결정. 'rotate'인 경우 emit 메소드가 rotate 이벤트를, 'alert'의 경우 alert 이벤트를 트리거.
- **심화 요구사항**
  - eslint에서 alert을 제한하는 옵션으로 에러가 발생
    - 해당 라인에 eslint-disable-line no-alert를 주석으로 추가하여 no-alert 옵션을 무효화
  - 버튼을 클릭할 때, 문자열과 클릭 횟수가 alert 창에 출력
    - 삼항 연산자를 사용하여 369 게임처럼 3의 배수는 'clap'이라는 문자열이, 그 외에는 클릭 횟수가 출력되도록 로직을 구현하였습니다.

## 특이사항

> 구현하면서 궁금했던 점이나, 아쉬웠던 부분이 있으면 적어주세요.
- **TDD 방법론 관련 질문**
  - TDD는 실패하는 테스트 코드 작성 → 통과하기 위한 최소한의 기능 구현 → 리팩토링인 걸로 이해하고 그렇게 개발을 진행했는데, 리팩토링 과정에서 테스트 코드를 불가피하게 수정하게 되었습니다. 예를 들어서 input 엘리먼트가 존재하는지 테스트 코드를 작성하고 → 이를 통과하기 위해 input 엘리먼트를 삽입하고 → 리팩토링 과정에서 input 엘리먼트를 커스텀 컴포넌트로 만들었는데, 이러면 테스트 코드가 Fail이 되어 다시 코드를 수정해야 했습니다. button 엘리먼트도 마찬가지였습니다.
애초에 두 번째 단계인 최소한의 기능 구현 과정에서 HTML 엘리먼트가 아닌 컴포넌트로 만드는 것이 바른 접근법이었는지, 강사님이었다면 어떤 방식으로 개발하셨을지 궁금합니다~!🙏🏻

## 제출 후 확인

* [x] 본인의 구현 결과를 `dev` 브랜치에 구현하고, `dev` 브랜치에서 `submission` 브랜치로 PR을 보내주세요.
* [x] Pull Request 생성 후 Reivewer에 @externship-master 와 @lkaybob 을 추가해주세요.

glowing713 committed:
chore: Add watch test npm script
Created watch-test script by adding --watchAll option
to test:unit script

glowing713 committed:
test: Testing input, button tag existance
문자열 입력란과 버튼 존재여부 확인하는 테스트코드 작성

glowing713 committed:
feat: Add input, button element to HelloWorld compnent
테스트를 통과하기 위한 최소한의 기능 추가

glowing713 committed:
chore: Change test lang to kor

glowing713 committed:
test: 요구사항 2번 테스트코드 작성
input에서 입력한 값은 실시간으로 반영되어
바로 밑에 표시가 되어야 한다.

glowing713 committed:
feat: 요구사항 2번 기능 구현
v-model 디렉티브를 이용하여 input과 p의 양방향
바인딩 구현

glowing713 committed:
refactor: Add base-input component
input 엘리먼트를 v-model과 v-on을 사용하여 컴포넌트화시킴.

glowing713 committed:
fix: Fixed test code to pass all cases
input 엘리먼트를 컴포넌트화 한 코드에 맞게 테스트코드 수정
input 엘리먼트를 찾던 방식 -> base-input 컴포넌트를 찾는 방식

glowing713 committed:
test: Add additional cases to testcase2
영어, 숫자, 한글 입력 케이스 추가

glowing713 committed:
refactor: Refactored test code
중복된 선언과 불필요한 체크 구문들 제거

glowing713 committed:
test: Add BaseInput component test code
emit event check, 콜백함수로 전달되는 인자가 올바른지 체크

glowing713 committed:
chore: Add modelValue prop

glowing713 committed:
test: Add test code of test case 3
버튼 클릭에 따라 텍스트가 왼쪽으로 한칸씩 이동하도록 테스트코드 작성

glowing713 committed:
test: Fixed test code of test case 3
테스트 코드 상에서 input 엘리먼트의 값을 불러오지 못하는 현상이 발생해서
두 번의 await 과정에서 setTimeout을 통해 문제를 해결했다고 생각했지만,
디버깅 과정에서 shallowMount를 사용하여 BaseInput 컴포넌트 내부의
input 엘리먼트를 불러오지 못한 것이 문제였다는 것을 발견하게 됨.
mount 함수로 변경하여 내부의 input 엘리먼트까지 트래킹하게 되었고,
올바른 테스트를 수행하는 코드로 수정할 수 있게 됨!

glowing713 committed:
feat: Add rotate function to button
버튼을 누르면 내부 메소드를 통하여 텍스트가 회전하게 되고,
변경사항이 input 엘리먼트에도 동시에 반영된다.

glowing713 committed:
refactor: Add base-button component
button 엘리먼트를 커스텀 컴포넌트화해서 재사용성을 높임
func prop으로 button의 기능을 지정해주고 컴포넌트 내부에서 prop에 따라
목적에 맞는 이벤트를 트리거해줌

glowing713 committed:
test: Add base-button test code
base-button 컴포넌트 테스트 코드 추가,
이에 맞게 HelloWorld 테스트 코드 일부 수정

glowing713 committed:
test: Add alert option case to base-button test code

glowing713 committed:
feat: Add alert emit event to basebutton component

glowing713 committed:
feat: 심화 요구사항 구현
알림 버튼을 클릭하면 입력 텍스트와 버튼 누른 횟수 출력.
추가로 횟수가 3의 배수이면 369 게임처럼 "clap"이 출력되도록 만듬!

k12ms26 commented on mission/src/components/HelloWorld.vue:
```diff
@@ -1,15 +1,40 @@
+      let template = `✍🏻 Input text: ${this.inputText} ✍🏻\n`;
+      this.btnPressed += 1;
+      template += (this.btnPressed % 3 ? `🔥 You just pressed ${this.btnPressed} times! 🔥` : '👏🏻👏🏻Clap👏🏻👏🏻');
+      alert(template); // eslint-disable-line no-alert
```
이 주석은 warning을 해결하지 못한 것에 대한 주석일까요?

glowing713 replied:
리뷰 감사드립니다:D eslint-disable-line 주석은 해당 라인만 eslint의 no-alert 룰을 해제하고자 사용하였습니다!
eslint 공식문서에 따르면 alert 구문을 제거하거나 커스텀 UI로 구현하라는 것을 확인할 수 있었습니다. 하지만 강사님께서 보여주신 데모와 디스코드에서 힌트를 주신 내용을 토대로 alert 함수를 사용해야겠다고 판단하게 되었습니다.
파일단위/프로젝트 단위로 no-alert 옵션을 부여할 수도 있겠지만, 규모가 커졌을 때 혹시나 의도치 않은 문제를 유발할까 싶어서 해당 라인만 옵션을 해제하는 방법을 선택했습니다!ㅎㅎ

k12ms26 reviewed (commented):
[좋은 점/잘한 점]
체계적인 테스트를 기반으로 코드를 짜신 것 같아 많은 참고가 될 것 같습니다

[의견제시]
코멘트 달아두었습니다 :)

2taesung commented:
[좋은 점/잘한 점]
- 차후 리팩토링을 고려한 코드들이 너무 깔끔하게 정리되어 있는 것 같습니다.
- 수정 사항이 바로 반영되는 코드를 추가해 효율을 높인 점
- BaseInput 컴포넌트의 Vue2 와 Vue3의 차이를 찝어 설명해주셔서 큰 도움이 됐습니다. 

[의견제시]
- 혹시 테스트 코드들을 모듈로 분리한 기준은 기능 별로 분리를 한 건가요 ???

daaeaad commented:
**[ 좋은점/잘한점 ]**
- 제공된 package.json을 수정하여 작업 효율성을 더 올린 점이 인상 깊었습니다.
- TDD방법론을 잘 적용하여 작업하신 것 같아 좋은 참고 자료가 될 것 같습니다.
- 두 가지의 버튼을 하나의 컴포넌트로 만들어 재사용성을 고려하신 점
- 블로그에 따로 학습내용을 정리하시는 부지런함 ...👍 배워야겠네요 ´, .̫ .`)

**[ 의견 제시 ]**
- 코멘트 남깁니다 :)

daaeaad commented on mission/src/components/HelloWorld.vue:
```diff
@@ -1,15 +1,40 @@
+      let template = `✍🏻 Input text: ${this.inputText} ✍🏻\n`;
+      this.btnPressed += 1;
+      template += (this.btnPressed % 3 ? `🔥 You just pressed ${this.btnPressed} times! 🔥` : '👏🏻👏🏻Clap👏🏻👏🏻');
+      alert(template); // eslint-disable-line no-alert
```
.eslintrc.js 파일의 rules에 'no-alrt': 'off'를 추가하여 설정을 변경하는 방법도 있더라구요 :)
저도 이 경고를 어떤 방법으로 해결할지 많이 고민해보았는데
혹시 알고 계셨다면 주석코드를 사용하신 특별한 이유가 있는지 궁금합니다.

(참고)
[https://stackoverflow.com/questions/35308336/eslint-disallow-use-of-alert-no-alert](url)

glowing713 replied:
리뷰 감사드립니다!!ㅎㅎ 오.. 저랑 같은 글을 보셨군요!
저도 eslint 공식문서와 링크로 공유해주신 글을 토대로 짧게 고민을 해보았었는데요,
프로젝트 자체에서 alert 창을 사용해야 하는 일이 많다면 파일단위/프로젝트단위로 해당 옵션을 해제하는 것도 좋을 것 같다고 생각했습니다.
그러나 혹시나 디버깅에 사용된 alert 창이 원치 않게 프로덕트 포함되어 배포되지는 않을까도 생각해보게 되었습니다,,,,
이런 문제를 예방하기 위해서는 eslint가 제공하는 옵션을 그대로 사용하고 저 라인에만 no-alert 옵션을 부여하는 것도 좋은 방법일 듯하여 한 줄 주석을 사용하게 되었습니다!

daaeaad replied:
아하.. 맞네요..! 저는 한번에 적용할 수 있는게 효율적이다 라는 생각만 하고 설정파일을 수정했는데...
거기까지 생각을 못하고 있었어요..!
선환님 코멘트 보고 크게 깨닫고 갑니다 👍 ..!!!!

glowing713 commented:
>

ohyein00 commented:
안녕하세요 4조에 조예인입니다!

- 재사용성을 고려한 컴포넌트를 고려하신 점이 인상깊었습니다
- 본인에게 맞는 세팅을 연구하신 것이 많은 자극이 되었습니다. 블로그 글 잘 보았습니다! 많은 공부가 되었어요😀
- TDD방법론이 잘 와닿지 않았는데 선환님의 코드를 보니 많은 생각을 하게 되었습니다 감사합니다!

ljunseok0514 commented:
- 심화기능 + 369 추가기능 까지 잘 구현하셨습니다!
- 컴포넌트를 2개로 나눠서 작업하셔서 더 깔끔한 것 같습니다.
- TDD 방법론을 잘 적용하신 것 같습니다.

glowing713 commented:
> * 혹시 테스트 코드들을 모듈로 분리한 기준은 기능 별로 분리를 한 건가요 ???

@2taesung 리뷰 감사드립니다!! 최대한 구현할 기능을 단위로 테스트 코드를 작성하려고 노력했던 것 같습니다,,ㅎㅎ
컴포넌트는 하나의 주된 기능만을 담당하도록 하고, 테스트 코드를 통해서 그 주된 기능이 무엇인지 표현하려는 방식으로 작성했습니다.
주된 로직이 구현되는 HelloWorld 컴포넌트의 테스트 코드는 미션으로 구현해야하는 기능별로 나누어서 작성해보았습니다!

lkaybob commented on mission/tests/unit/HelloWorld.spec.js:
```diff
@@ -0,0 +1,45 @@
+
+describe('HelloWorld.vue', () => {
+  const wrapper = mount(HelloWorld);
+  const baseInput = wrapper.findComponent({ name: 'base-input' });
```
질문주셨던 내용(refactor할 때 테스트가 깨진다) 관련해서, base-input을 찾는 것보다는 `button.base-button`을 찾으셨으면 컴포넌트화해도 깨지지 않았을 겁니다.
* button 또한 어느 수준에서 보면 implementation level이기 때문에 테스트용 attribute를 따로 만들어서 검증하시는 것이 더 좋은 방법이긴 할겁니다.

glowing713 replied:
늦은 시간까지 리뷰 감사드립니다:D
엘리먼트에 테스트 코드용 attribute를 추가하는 것이 코드 가독성을 해치지는 않을까 하는 생각에 지양하려고 했던 것 같습니다,,
하지만 말씀해주신 것처럼 코드의 구현이 바뀌어도 테스트 결과를 유지하면서 빠르고 정확하게 개발해나가기 위해서 적용해보아야겠다는 생각이 듭니다! 다음 과제부터는 이 방식으로 접근해서 테스트 코드를 작성해보겠습니다!ㅎㅎ

lkaybob commented on mission/tests/unit/HelloWorld.spec.js:
```diff
@@ -0,0 +1,45 @@
+    await baseInput.setValue('123123112');
+    expect(textPrint.text()).toBe('123123112');
+
+    await baseInput.setValue('이건 한글 버전의 테스트 케이스');
```
특수한 예외상황이 아니라면 tc당 expect를 통한 assertion은 최소화하는 것이 좋습니다.

lkaybob commented on mission/tests/unit/BaseInput.spec.js:
```diff
@@ -0,0 +1,34 @@
+  });
+
+  it('emit 이벤트의 콜백함수 인자로 전달되는 값이 입력과 동일해야 한다', async () => {
+    // await inpElement.setValue('hello!');
```
사용되지 않는 주석들은 미리 지워주세요

lkaybob commented on mission/tests/unit/BaseInput.spec.js:
```diff
@@ -0,0 +1,34 @@
+    // await inpElement.setValue('hola!');
+    inpElement.element.value = 'hello!';
+    inpElement.trigger('input');
+    expect(wrapper.emitted('update:modelValue')[1][0]).toEqual('hello!');
```
1(혹은 2,3)과 0은 magic number로 보입니다. 상수/변수로 선언을 하시는 것을 권장드립니다.

lkaybob commented on mission/tests/unit/BaseButton.spec.js:
```diff
@@ -0,0 +1,22 @@
+import BaseButton from '@/components/BaseButton.vue';
+
+describe('BaseButton.vue', () => {
+  const wrapper = mount(BaseButton);
```
wrapper 선언시에는 빈 값으로 두고 beforeEach를 이용해서 mount를 하시는 것을 권장드립니다.
* beforeEach를 사용하면 각 테스트 케이스는 서로 독립적인 VueWrapper로 테스트할 수 있습니다.

glowing713 replied:
다른 수강생분들 코드와 강사님께서 남겨주신 리뷰들을 토대로 beforeEach를 꼭 적용해보아야겠다고 생각했습니다! 감사합니다:)

lkaybob reviewed (commented):
TDD 방법론을 지키시면서 테스트 코드를 먼저 작성하셨고, 테스트 케이스들도 잘 작성해주셨습니다. 또한 변수/메서드 명들이 명시적이어서 이해하기 편했습니다.

기본적으로 테스트 코드는 리팩토링을 해도 깨지지 않아야 하는 것이 best practice입니다. 질문주신 부분에 대해서는 line comment로 의견드렸으니 한 번 확인해주세요.

> The rule of thumb is that a test should not break on a refactor, that is, when we change its internal implementation without changing its behavior. If that happens, the test might rely on implementation details.

glowing713 committed:
test: Change component find method
get 메소드를 사용하여 엘리먼트가 존재하지 않으면 에러를 띄우도록 설정
testing attribute를 사용하여 엘리먼트를 직접 찾도록 설정
(+ vscode가 vue-cli 프로젝트의 src경로 alias를 잘 인식하도록 jsconfig 수정)

glowing713 committed:
refactor: Add data-test attribute to DOM elements
변경된 테스트코드를 통과할 수 있도록 찾고자하는 엘리먼트에 data-test 속성 추가

glowing713 committed:
refactor: Reduce assertion in tc
한 tc에 불필요하게 많은 assertion이 존재하지 않도록 수정

glowing713 committed:
refactor: Remove unnecessary comments

glowing713 committed:
test: Add beforeEach and find DOM elmt by data-test att
wrapper 객체와 element를 beforeEach 내부에 선언
엘리먼트 찾아오는 방식을 data-test를 사용하도록 변경

glowing713 committed:
refactor: Add emitParameter variable
wrapper.emitted()의 출력값에 배열 인덱스로 접근하는 구문의 의미가 모호하므로
출력값이 의미하는 바를 명시하는 변수에 할당

glowing713 committed:
test: Add beforeEach method
